### PR TITLE
[13.0][ADD] base_menu_visibility_restriction: Restrict (with groups) menu visibilty

### PR DESCRIPTION
**Pull request by victoralmau**
This addon adds "Restricted groups" in Menu items.

Please @pedrobaeza, @joao-p-marques and @Yajo review it.
@Tecnativa TT26721

PD: Need to depends `web_tour` to prevent error in test ("Missing web_tour model").

### CODE REVIEW

File: base_menu_visibility_restriction/readme/USAGE.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+To use this module, you need to go to:
+
+* Activate the developer mode
+* Settings > Users & Companies > Groups and create new record with "Name": Hide menu items custom and "Users": Admin
```
Comment (joao-p-marques) [inline base_menu_visibility_restriction/readme/USAGE.rst]: I would change this to something that could be more clear to the user, like:
```suggestion
* Settings > Users & Companies > Groups and create new record(s) with "Name" of your choice and add to "Users" the ones you want to be able to access a certain Menu Item 
```

File: base_menu_visibility_restriction/readme/USAGE.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+* Settings > Technical > User interface > Menu Items, search "Apps" record and update "Excluded groups": Hide menu items custom.
```
Comment (joao-p-marques) [inline base_menu_visibility_restriction/readme/USAGE.rst]:
```suggestion
* Settings > Technical > User interface > Menu Items, search "Apps" record and update "Excluded groups"  with the one(s) you created earlier.
```

**Review by joao-p-marques — APPROVED** [review]
Code and functional review :+1:. Good job!

Just some suggestions to improve ease of use to new users. 

Also, seems like you forgot to generate the README.

**Comment by victoralmau** [comment]
> Code and functional review +1. Good job!
> 
> Just some suggestions to improve ease of use to new users.
> 
> Also, seems like you forgot to generate the README.

Thanks Joao. Readme file generate in merge with ocabot, right?

File: base_menu_visibility_restriction/readme/DESCRIPTION.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+This addon adds "Restricted groups" in Menu items.
```
Comment (yostashiro) [inline base_menu_visibility_restriction/readme/DESCRIPTION.rst]: An alternative text suggestion, if it's clearer:
This addon lets you assign "excluded groups" to menu items. If a user belongs to a group that is assigned to a menu item as an excluded group, the user will not be able to see the menu item.

File: base_menu_visibility_restriction/readme/USAGE.rst [hunk 4]
```diff
@@ -0,0 +1,6 @@
+To use this module, you need to go to:
+
+* Activate the developer mode
+* Settings > Users & Companies > Groups and create new record(s) with "Name" of your choice and add to "Users" the ones you want to be able to access a certain Menu Item
+* Settings > Technical > User interface > Menu Items, search "Apps" record and update "Excluded groups"  with the one(s) you created earlier.
```
Comment (kittiu) [inline base_menu_visibility_restriction/readme/USAGE.rst]: Just that, on the first time reading this, I don't know what means "Apps". ;) 

This is a useful module.

```suggestion

Given you want to restrict the "Apps" menu,

* Settings > Technical > User interface > Menu Items, search "Apps" record and update "Excluded groups"  with the one(s) you created earlier.
```

**Comment by kittiu** [comment]
@victoralmau I wonder if this module can be extended for Print, Action menu easily? Just asking.

**Review by kittiu — APPROVED** [review]

**Comment by victoralmau** [comment]
> @victoralmau I wonder if this module can be extended for Print, Action menu easily? Just asking.

This addon should not add that functionality (because it is intended only to limit the menu) but IF it would be possible to make another "similar" (that the purpose is the same) to Print or other actions.

File: base_menu_visibility_restriction/models/ir_ui_menu.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        column1="menu_id",
+        column2="gid",
+        string="Excluded Groups",
+    )
+
+    @api.model
+    @tools.ormcache("frozenset(self.env.user.groups_id.ids)", "debug")
+    def _visible_menu_ids(self, debug=False):
+        """ Return the ids of the menu items visible to the user. """
+        # retrieve all menus, and determine which ones are visible
```
Comment (joao-p-marques) [inline base_menu_visibility_restriction/models/ir_ui_menu.py]: @victoralmau I have just noticed you are rewritting this method without calling the super method. That can break some functionallity. Is it really necessary?

Comment (victoralmau) [reply base_menu_visibility_restriction/models/ir_ui_menu.py]: Super improved and simplify function.

File: base_menu_visibility_restriction/models/ir_ui_menu.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2020 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from odoo import api, fields, models, tools
+
+
+class IrUiMenu(models.Model):
+
+    _inherit = "ir.ui.menu"
+
+    excluded_groups_id = fields.Many2many(
```
Comment (pedrobaeza) [inline base_menu_visibility_restriction/models/ir_ui_menu.py]: Field name is not correct according guidelines:
```suggestion
    excluded_group_ids = fields.Many2many(
```
Maybe you have selected seeing other existing fields that are incorrect in these models, but we shouldn't extend that bad names on new ones.

File: base_menu_visibility_restriction/readme/USAGE.rst [hunk 7]
```diff
@@ -0,0 +1,6 @@
+To use this module, you need to go to:
```
Comment (pedrobaeza) [inline base_menu_visibility_restriction/readme/USAGE.rst]:
```suggestion
To use this module, you need to:
```

File: base_menu_visibility_restriction/readme/USAGE.rst [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+
+* Activate the developer mode
+* Settings > Users & Companies > Groups and create new record(s) with "Name" of your choice and add to "Users" the ones you want to be able to access a certain Menu Item
+* Settings > Technical > User interface > Menu Items, search "Apps" record and update "Excluded groups"  with the one(s) you created earlier.
+* Login as demo user and check that "Apps" menu item not show.
```
Comment (pedrobaeza) [inline base_menu_visibility_restriction/readme/USAGE.rst]: Several things:

* You have to put generic instructions, not a specific case. You can afterwards specify an example.
* Each step is a point. Don't group several of them together.
* Use numbered bullets.
* The group creation is out of the scope of this module.
```suggestion
#. Go to *Settings > Technical > User interface > Menu Items*.
#. Search for any menu and edit it.
#. Update "Excluded groups" with one group.
#. Login with a user of that group, and you won't see such menu.

You can try with demo data for the menu X and user demo.
```
And add such demo data :wink:

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by victoralmau** [comment]
Travis is green now.

File: base_menu_visibility_restriction/readme/USAGE.rst [hunk 9]
```diff
@@ -0,0 +1,10 @@
+To use this module, you need to:
+
+* Activate the developer mode
```
Comment (pedrobaeza) [inline base_menu_visibility_restriction/readme/USAGE.rst]: Why putting this with different bullet ?

File: base_menu_visibility_restriction/demo/res_group.xml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo noupdate="1">
+    <record id="group_demo_user" model="res.groups">
+        <field name="name">Demo user</field>
```
Comment (pedrobaeza) [inline base_menu_visibility_restriction/demo/res_group.xml]: Give a more explanatory group name

Comment (pedrobaeza) [reply base_menu_visibility_restriction/demo/res_group.xml]: I mean to give an explanatory name to this group, not to put a comment:
```suggestion
        <field name="name">Menu visibility restriction demo group</field>
```

**Review by pedrobaeza — APPROVED** [review]

**Review by HaraldPanten — APPROVED** [review]

**Pull request merged**